vicodyn/src/vicodyn/proxy.cpp

```diff
@@ -336,112 +364,98 @@ class vicodyn_dispatch_t : public std::enable_shared_from_this<vicodyn_dispatch_
             .execute(std::bind(&vicodyn_dispatch_t::on_backward_error, this, ph::_1, ph::_2, ph::_3));
 
         backward_dispatch.on_discard([&](const std::error_code& ec) {
-            try {
-                backward_stream.error({}, make_error_code(vicodyn_errors::upstream_disconnected),
-                                      "vicodyn upstream has been disconnected");
-            } catch (const std::exception& e) {
-                COCAINE_LOG_WARNING(endpoint.logger, "could not send error {} to upstream - {}", ec, e);
-            }
+            client_stream.error({}, ec, "peer has been disconnected");
         });
     }
 
-    ~vicodyn_dispatch_t() {
-    }
-
     auto enqueue(const hpack::headers_t& headers, const std::string& event) -> std::shared_ptr<dispatch<app_tag>> {
-        return mutex.apply([&](){
-            return enqueue_unsafe(headers, event);
-        });
+        unsafe_.buffer.store_event(headers, event);
+        if (!choose_endpoint(headers, event)) {
```

Comment:
From this row exception still can be raised.
But same behavior was before: https://github.com/interiorem/cocaine-plugins/pull/15
Fix of exception from enqueue will be done in COCAINE-2656.